cmd/keytransparency-delegate/main.go

```diff
@@ -0,0 +1,117 @@
+	"net/http"
+
+	"github.com/google/keytransparency/cmd/serverutil"
+	"github.com/google/keytransparency/core/managementserver"
+	"github.com/google/keytransparency/impl/sql/engine"
+	"github.com/google/keytransparency/impl/sql/keysets"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/reflection"
+
+	pb "github.com/google/keytransparency/core/api/usermanager/v1/usermanager_proto"
+	_ "github.com/google/trillian/crypto/keys/der/proto"
+	grpc_prometheus "github.com/grpc-ecosystem/go-grpc-prometheus"
+)
+
+var (
+	addr         = flag.String("addr", ":8080", "The ip:port combination to listen on")
+	metricsAddr  = flag.String("metrics-addr", ":8081", "The ip:port to publish metrics on")
+	serverDBPath = flag.String("db", "test:zaphod@tcp(localhost:3306)/test", "Database connection string")
+	keyFile      = flag.String("tls-key", "genfiles/server.key", "TLS private key file")
+	certFile     = flag.String("tls-cert", "genfiles/server.crt", "TLS cert file")
+
+	instance = flag.Int64("instance", 0, "Instance number. Typically 0.")
+)
+
+func openDB() *sql.DB {
+	db, err := sql.Open(engine.DriverName, *serverDBPath)
+	if err != nil {
+		glog.Exitf("sql.Open(): %v", err)
+	}
+	if err := db.Ping(); err != nil {
+		glog.Exitf("db.Ping(): %v", err)
+	}
+	return db
+}
+
+func main() {
+	flag.Parse()
+
+	// Open Resources.
+	sqldb := openDB()
+	defer sqldb.Close()
+
+	keysetdb, err := keysets.New(sqldb)
+	if err != nil {
+		glog.Exitf("Failed to create keyset table: %v", err)
+	}
+
+	creds, err := credentials.NewServerTLSFromFile(*certFile, *keyFile)
+	if err != nil {
+		glog.Exitf("Failed to load server credentials %v", err)
+	}
+
+	// Create gRPC server.
+	svr := managementserver.New(*instance, keysetdb)
+	grpcServer := grpc.NewServer(
+		grpc.Creds(creds),
+		grpc.StreamInterceptor(grpc_prometheus.StreamServerInterceptor),
+		grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor),
+	)
+	pb.RegisterUserManagerServiceServer(grpcServer, svr)
+	reflection.Register(grpcServer)
+	grpc_prometheus.Register(grpcServer)
+
+	// Create HTTP handlers and gRPC gateway.
+	tcreds, err := credentials.NewClientTLSFromFile(*certFile, "")
+	if err != nil {
+		glog.Exitf("Failed opening cert file %v: %v", *certFile, err)
+	}
+	gwmux, err := serverutil.GrpcGatewayMux(*addr, tcreds,
+		pb.RegisterUserManagerServiceHandlerFromEndpoint)
+	if err != nil {
+		glog.Exitf("Failed setting up REST proxy: %v", err)
+	}
+
+	// Insert handlers for other http paths here.
+	mux := http.NewServeMux()
+	mux.Handle("/", gwmux)
+
+	metricMux := http.NewServeMux()
+	metricMux.Handle("/metrics", promhttp.Handler())
+	go func() {
+		log.Printf("Hosting metrics on %v", *metricsAddr)
+		if err := http.ListenAndServe(*metricsAddr, metricMux); err != nil {
+			log.Fatalf("ListenAndServeTLS(%v): %v", *metricsAddr, err)
+		}
+	}()
+	// Serve HTTP2 server over TLS.
```

Comment:
HTTP2? Is that a typo?
Nit on wording:  `// Serve HTTP over TLS.`

Reply:
It's probably a typo, but it turns out it's also technically correct.
gRPC only happens over HTTP2, which has a bunch of multiplexing magic.